### DISCUSSION TITLE
chapter_tree/array_representation_of_tree/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#472] [Announcements]
# chapter_tree/array_representation_of_tree/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_tree/array_representation_of_tree/

**Comment by fallenleavesguy** [comment]
讲得这么好为什么没有评论

**Reply by krahets** [reply]
哈哈，可能是因为这一节是最近单独分出来的

**Comment by Focu230104** [comment]
"完全二叉树非常适合使用数组来表示。回顾完全二叉树的定义，\(\text{None}\) 只出现在最底层且靠右的位置"，是靠左吧，大佬是不是误笔了

**Reply by krahets** [reply]
原文没问题，在完全二叉树的最底层中，节点靠左、None 靠右

**Reply by Focu230104** [reply]
不好意思，看错了，以为是节点，谢谢大佬

**Comment by xiaoyeaichif** [comment]
不仅图画的好，讲的也很好。作为回顾和复习真的太棒啦！如果能附上代码就更美了！！！

**Reply by krahets** [reply]
感谢建议，这一节的代码稍后加上～

**Reply by xiaoyeaichif** [reply]
感谢回复，希望越做越好！对于新手太友好了这网站！

**Comment by LongLiveY96** [comment]
```java
if (order == "pre")
    res.add(val(i));
dfs(left(i), order, res);
// 中序遍历
if (order == "in")
    res.add(val(i));
dfs(right(i), order, res);
// 后序遍历
if (order == "post")
```
这里的字符串对比是不是有问题啊，不应该用"xxx".equals(order)吗

**Reply by krahets** [reply]
在 Java 中，对于基本数据类型，`==` 用于对比两个变量的值是否相等。

对于引用类型，两种符号的工作原理不同：

- `==` ：用来比较两个变量是否指向同一个对象，即它们在内存中的位置是否相同。
- `equals()`：用来对比两个对象的值是否相等。

因此如果要对比值，我们一般会用 `equals()` 。然而这里的 "pre" , "in" , "post" 都存储在字符串常量池中，等式左右两个 "pre" 指向同一个对象，因此可以用 `==` 来比较。

**Comment by zhuasdfg** [comment]
妙！妙不可言！前中后序遍历实现只是res.add()位置不同，递归实在是太牛了。

**Comment by hazukieq** [comment]
7.3.2 小节中 C语言实现的代码有点奇怪，请问 C语言中有 `vector` 向量这个关键字吗？我仔细瞧了瞧，代码中也没有用宏定义 vector。我在 cppreference 网站中的 C keywords 中也没有找到 `vector`。
![C keywords shortcut ](https://www.hazukieq.top/imgs/Ckeywords.png)

**Reply by ahnovice** [reply]
我也没看懂😭

**Comment by AY-Front** [comment]
大佬，有个小小的建议，要是代码里附上一些测试例子对小白就更友好了

**Reply by krahets** [reply]
源码里都有测试样例呀，例如 [Python 代码 ](https://github.com/krahets/hello-algo/blob/0a2ad4489a8c103b24a800892f83de1b6644b5cd/codes/python/chapter_tree/array_binary_tree.py#L88-L118) 。

**Comment by topul** [comment]
这一节太有用了，之前刷leetcode，看到二叉树的数组表示就感觉有点懵，看完讲解感觉舒服多了

**Comment by i-Remus** [comment]
`class ArrayBinaryTree` 的 `__dfs`方法写的太美味了！！

**Reply by wpf19950321** [reply]
但是迷惑性很强哈哈

**Comment by carrot-butcher** [comment]
作者你好，关于 __dfs 这段代码我有点没看懂。

```py
def __dfs(self, i: int, order: str):
        """深度优先遍历"""
        if self.val(i) is None:
            return
        # 前序遍历
        if order == "pre":
            self.res.append(self.val(i))
        self.__dfs(self.left(i), order)
        # 中序遍历
        if order == "in":
            self.res.append(self.val(i))
        self.__dfs(self.right(i), order)
        # 后序遍历
        if order == "post":
            self.res.append(self.val(i))
```

就比如说：如果是前序遍历，那self.__dfs(self.left(i), order)往二叉树最深处递完后，就只是遍历了二叉树最左侧的一支，那相应的这一分支的右节点和根节点怎么遍历的呢？好像这个函数里面也没有关于parent函数的应用。
链表前序遍历二叉树，我记得是“根->左->右”，就是如下（上节内容）：

```py
def pre_order(root: TreeNode | None):
""" 前序遍历"""
    if root is None:
       return
    res.append(root.val)
    pre_order(root=root.left)
    pre_order(root=root.right)
```

为什么数组的二叉树遍历里有部分代码不用写呢？希望作者能帮我解答这个问题。

**Reply by krahets** [reply]
Hi ，如果我们只看前序遍历，就可以将中序、后序遍历代码删除掉，函数变为：

```python
def __dfs(self, i: int, order: str):
        """深度优先遍历"""
        if self.val(i) is None:
            return
        self.res.append(self.val(i))
        self.__dfs(self.left(i), order)
        self.__dfs(self.right(i), order)
```

这样看是否就和上节的前序遍历代码一致了？不同点仅在于获取值、左子节点、右子节点的方式不同。

**Reply by lihuxiao1** [reply]
是我没看明白嘛，我看那个代码里的pre 部分，只有self.__dfs(self.left(i), order)，而作者你回答的是有两行self.__dfs(self.left(i), order)
self.__dfs(self.right(i), order)

**Reply by Hadias94** [reply]
看了半天看出来的，那个递归的动作不在if语句中

**Reply by krahets** [reply]
展开这样看，更清楚一些：

```python
def __dfs(self, i: int, order: str):
    """深度优先遍历"""
    if self.val(i) is None:
        return
    if order == "pre": self.res.append(self.val(i)) # 前序遍历
    self.__dfs(self.left(i), order) # 递归左子树
    if order == "in": self.res.append(self.val(i)) # 中序遍历
    self.__dfs(self.right(i), order) # 递归右子树
    if order == "post": self.res.append(self.val(i)) # 后序遍历
```

**Reply by Verlog123** [reply]
注意看缩进，眼睛都给我看花了，原来是缩进没注意看，self.dfs(self.left(i), order)和self.dfs(self.right(i), order)这2行代码并不在if循环语句里面，而是在外面

**Comment by Leezgion** [comment]
请问上例JS代码中
```javascript
  /* 获取索引为 i 节点的父节点的索引 */
    parent(i) {
        return (i - 1) / 2;
    }
```
```javascript
问题描述 举例
let tree = [1, 2, 3, 4, null, 6, 7, 8, 9, null, null, 12, null, null, 15];
parant(3) //------>return 1
parant(4) //------>return 1.5
```
> 请问数组的索引值不应该为`整数`吗，如果为整数那么上述parent()中 return 返回的值不应该进行向下取整吗？

**Reply by github15238484509** [reply]
> 我想明白了，
1.  一个节点分为两个孩子，比如 节点为1 的时候，左孩子：1 * 2 +1  =3；右孩子 1 * 2 + 2  = 4
2. 用左推父的时候需要 -1 ； 右推父的时候 -2 ，但是作者只用了 左推父，没有想到 右推父的情况；向下取整就好了

**Reply by kaka59** [reply]
我也有同样的疑问，个人认为应该为
`return Math.floor(  (i - 1) / 2 );`

**Reply by superFlyMoose** [reply]
右推父也成立，因为这是整数除以整数，本就可以向下取整

**Comment by showlibia** [comment]
DAY6

**Comment by blxzer77** [comment]
```c++
/* 深度优先遍历 */
void dfs(int i, string order, vector<int> &res) {
    // 若为空位，则返回
    if (val(i) == INT_MAX)
        return;
    // 前序遍历
    if (order == "pre")
        res.push_back(val(i));
    dfs(left(i), order, res);
    // 中序遍历
    if (order == "in")
        res.push_back(val(i));
    dfs(right(i), order, res);
    // 后序遍历
    if (order == "post")
        res.push_back(val(i));
}
```

研究了好久才发现递归不在if语句内 笨死自己

**Reply by amoyonglee** [reply]
谢谢你的留言! 我一直没有看明白这段代码, 经过你的提醒, 我终于明白啦!

**Reply by Lyrilon** [reply]
我也和你犯了一样的错误www，感谢你的指出。

**Reply by dongpingwang** [reply]
尼玛，这代码太不规范了

**Reply by 1037861070** [reply]
因为if方法体只有一行是不用中括号的，省略了{}，但是编码规范是只要出现if的地方，方法体最好都加上{}，这里又是三种遍历都复用了部分代码，所以有些难理解

**Comment by zoujia** [comment]
作者你好，获取 左/右 子节点索引的方法，获取到的索引可能是超出范围的，比如索引 8 ，左、右子节点索引分别为 17、18，已超出最大 14 了~
能不能这么理解：实际应用中，单纯获取索引没有太大意义，必须配合 val(i) 来获取相应位置的数据，因为 val(i) 已经校验索引边界了~

**Reply by krahets** [reply]
Hi，是的～当然也可以在获取子节点索引的方法中加上校验，具体看实际需求

**Comment by 1515292252** [comment]
动态图在哪里  只有像PPT一样的静态图？

**Reply by krahets** [reply]
是的～

**Comment by selear** [comment]
```java
 /* 深度优先遍历 */
private void dfs(Integer i, String order, List<Integer> res) {
    // 若为空位，则返回
    if (val(i) == null)
        return;
    // 前序遍历
    if (order == "pre")
        res.add(val(i));
    dfs(left(i), order, res);
    // 中序遍历
    if (order == "in")
        res.add(val(i));
    dfs(right(i), order, res);
    // 后序遍历
    if (order == "post")
        res.add(val(i));
    }
```
以上截取自`array_binary_tree.java`，Java代码两个字符串内容的比较应当使用equals而非`==`，`==`比较的是两个对象所指向的地址是否相同的。

换句话说，这里的代码应当为：
```java
 /* 深度优先遍历 */
private void dfs(Integer i, String order, List<Integer> res) {
    // 若为空位，则返回
    if (val(i) == null)
        return;
    // 前序遍历
    if ("pre".equals(order))
        res.add(val(i));
    dfs(left(i), order, res);
    // 中序遍历
    if ("in".equals(order))
        res.add(val(i));
    dfs(right(i), order, res);
    // 后序遍历
    if ("post".equals(order))
        res.add(val(i));
    }
```

**Reply by krahets** [reply]
谢谢指正！这里虽然不影响结果的正确性（由于字符串常量池机制，它们的地址也是相通的），但用 equals 更加规范

**Comment by unicorn-option** [comment]
数组表示的Swift代码里面获取左右子节点索引是不是没有return

**Reply by nuomi1** [reply]
单行表达式可以省略 return。

**Comment by ZChenn58** [comment]
作者你好，关于这段代码，我有点没看懂。
```py
def __dfs(self, i: int, order: str):
        """深度优先遍历"""
        if self.val(i) is None:
            return
        # 前序遍历
        if order == "pre":
            self.res.append(self.val(i))
        self.__dfs(self.left(i), order)
        # 中序遍历
        if order == "in":
            self.res.append(self.val(i))
        self.__dfs(self.right(i), order)
        # 后序遍历
        if order == "post":
            self.res.append(self.val(i))
```
为什么中序遍历是`self.__dfs(self.right(i), order)`呢？为什么是右？

**Reply by hpstory** [reply]
Hi @Derry-celly ，因为中序遍历的顺序是左中右，dfs函数可执行的部分变成
```
def dfs(self, i: int, order: str):
    """深度优先遍历"""
    if self.val(i) is None:
        return
    self.dfs(self.left(i), order)
    self.res.append(self.val(i))
    self.dfs(self.right(i), order)
```
所以当前节点访问后应该访问右子节点

**Reply by ZChenn58** [reply]
理解了！感谢你的及时回复！

**Comment by HangSeven** [comment]
```java
  /* 获取索引为 i 节点的父节点的索引 */
    public Integer parent(int i) {
        return (i - 1) / 2;
    }
```
怎么确定这个节点不是右子节点，不是（i - 2）/ 2

**Reply by HangSeven** [reply]
明白了，整数除以整数还是整数刚好包括了，－永远是对的

**Comment by bnufw** [comment]
`def dfs(self, i: int, order: str)`
这个方法是不是不能直接调用啊，因为self.res在这个方法里没有定义？

**Reply by krahets** [reply]
Hi，我们不会直接调用它，它是前序、中序、后序遍历的工具方法

**Comment by 2lin123** [comment]
你好，这段代码
```cpp
/* 节点数量 */
    int size() {
        return tree.size();
    }
```
不应该是返回数组的大小吗，如果有空位的话，就不应该是节点的数量吧

**Reply by krahets** [reply]
谢谢指正，这里应为“数组容量”

**Comment by llhznb** [comment]
这三种遍历，能够完全遍历吗？比如前序遍历，一直找左子树，当索引超量程后return，右边的数也遍历了吗？

**Reply by llhznb** [reply]
擦，看错了，后面递归不在if里面

**Reply by KeeVor** [reply]
哈哈哈，感谢你提醒，我本来也有这个疑问呢，回头一看确实dfs递归调用都不在if里面。

**Reply by Aozaky** [reply]
是的！这段代码非常简洁与巧妙！

**Reply by LuoShuip** [reply]
感谢

**Comment by Clarkszw** [comment]
图7-13, 树上的灰色数组索引是不是标错了?

**Reply by curly210102** [reply]
应该是故意标错的，来表达非完全二叉树中的节点索引和数组索引不对应，非完全二叉树无法用不带 None 的数组来表示

**Reply by krahets** [reply]
@curly210102 补充：该数组对应的二叉树结构不唯一

**Comment by qw124341789** [comment]
你好，作者看完这章有个问题想问一下，用以下代码进行  前，中，后序遍历，应该没有特殊情况而发生无法预料的意外吧

**Reply by qw124341789** [reply]
```
void pre(Titree *tree){
	if (tree == NULL){ return; }
	printf("%d", tree->data);
	pre(tree->lchild);
	pre(tree->rchild);
}//前序遍历
void mid(Titree *tree){
	if (tree == NULL){return;}
	mid(tree->lchild);
	printf("%d", tree->data);
	mid(tree->rchild);
}//中序遍历
void post(Titree *tree){
	if (tree == NULL){ return; }
	post(tree->lchild);
	post(tree->rchild);
	printf("%d", tree->data);
}//后序遍历
```

**Reply by qw124341789** [reply]
完全二叉树用这段代码走了一遍，可以完美实现功能。

**Comment by KITSUNE234** [comment]
作为新手，dfs函数的写法感觉真的很巧妙，复用性很高啊

**Comment by chaojiwudidashuaige492** [comment]
我觉得 “映射公式的角色相当于节点中的引用” 会更好一点

**Reply by krahets** [reply]
谢谢建议！

**Comment by Magicperformer** [comment]
作者你好~请问7.3.3的二叉树局限性中，“增删节点需要通过数组插入与删除操作实现，效率较低。”这一点我不是非常可以理解，是说时间复杂度会到达O(n)的效率低吗？

**Reply by Magicperformer** [reply]
因为如果想通过数组来唯一表达一棵二叉树，那么这棵二叉树应该是一棵满二叉树，又或者是在数组中填充None代表空节点的常规二叉树，对于这样的数组的插入和删除应该不会导致数组中插入和删除带来的整体元素挪移（O(n)复杂度），因为只需要将数组对应的元素置换为None或者赋值就可以，这样子二叉树的插入和删除应该是O(1)的复杂度呀？请问我的理解对吗

**Reply by wbzyo** [reply]
如果直接将元素设为None，那理论上这个结点没有任何子节点了，就会出现没有父节点的两个子节点，那这也不能称为一棵二叉树了

**Reply by 1435646097** [reply]
但是就算是链表表示的二叉树，父节点灰删除了，子节点应该也都要一并删除吧。

**Comment by CJSen** [comment]
```python
def parent(self, i: int) -> int | None:
        """获取索引为 i 节点的父节点的索引"""
        return (i - 1) // 2
```
这边是不是应该考虑一下，如果i==0的情况，加个判断语句比较好。
```python
def parent(self, i: int) -> int | None:
        """获取索引为 i 节点的父节点的索引"""
        if i == 0:
            return None
        return (i - 1) // 2
```

**Reply by krahets** [reply]
在这个类的设计中，边界判断是在 `val()` 方法中实现的。因此我们没有在 `parent()`, `left()`, `right()` 中做边界判断

**Comment by nzb15555196162** [comment]
代码有问题吧,

修改后如下代码
```java
//    深度优先遍历
    public void dfs(int i, String order, List<Integer> res) {
//        若为空位，则返回
        if(val(i) == -1) {
            return;
        }
//        前序遍历
        if("pre".equals(order)) {
            res.add(val(i));
//            根节点 -> 左子树 -> 右子树
            dfs(left(i), order, res);
            dfs(right(i), order, res);
        }
//        中序遍历
        if("in".equals(order)) {
//            访问优先级：左子树 -> 根节点 -> 右子树
            dfs(left(i), order, res);
            res.add(val(i));
            dfs(right(i), order, res);
        }
//        后序遍历
        if("post".equals(order)) {
//            访问优先级：左子树 -> 右子树 -> 根节点
            dfs(left(i), order, res);
            dfs(right(i), order, res);
            res.add(val(i));
        }
    }
```

**Reply by KLYkl** [reply]
我认为原代码没问题。就是你后面那一段if的精简版而已，每个if都是根据输入的order判断运不运行，我把if判断为假的去掉，如果order是pre，原代码就是：
```python
res.add(val(i));
dfs(left(i), order, res);
dfs(right(i), order, res);
```
order等于in就是
```python
dfs(left(i), order, res);
res.add(val(i));
dfs(right(i), order, res);
```
后序也一样，就是跟之前的递归一模一样。还有就是你把if(val(i) == null) 改成if(val(i) == -1) ，原本数组里面可以存null的。

**Reply by nzb15555196162** [reply]
了解了，这里有个建议是if/else加上括号,不然很容易让人迷惑，另外这里if(val(i) == -1)这里我自己val(int x)返回int

**Reply by pianfan** [reply]
原代码dfs方法既精简又直接，换成以下格式也许会让你更好理解：
```cpp
void dfs(int i, string order, vector<int> &res) {
			if (val(i) == INT_MAX) return;
				
			if (order == "pre") res.push_back(val(i));
			dfs(left(i), order, res);
			if (order == "in") res.push_back(val(i));
			dfs(right(i), order, res);
			if (order == "post") res.push_back(val(i));
		}

```
其本质就在于，三个深度优先遍历都是先访问左节点再访问右节点，只是访问“根”节点的时机不同罢了

**Comment by YoRHa-Arya** [comment]
C# 中 DFS() 好巧妙，原来还能这样把深度遍历写成一个函数，学到了

**Reply by sishugang** [reply]
是的，看了两遍才看懂，真的好简洁，C++代码也是一样的

**Comment by Yojay97** [comment]
感觉二叉树的专业概念好多，看的时候需要截一个图放在旁白- -

**Reply by jiuerya7** [reply]
用用Snipaste，免费的截图软件，很好用

**Reply by nanmingjushi** [reply]
用 https://pixpin.cn/

**Comment by hyxenbug** [comment]
可以更详细说明一下为什么“增删节点需要通过数组插入与删除操作实现，效率较低。”吗？
是因为增加节点时可能要扩容，删除节点需要把子节点全都变成None吗？索引的映射公式已经把有可能存在节点的位置都安排好了，感觉应该不需要像数组插入与删除操作时移动后面的元素吧？

**Reply by YJExpand** [reply]
可以去看第4章，就知道为什么数组的插入和删除的效率低了

**Reply by jiuerya7** [reply]
因为需要如果添加/删除，你需要维护索引关系：
2i+1，2i+2，就需要移动插入/删除位置节点之后的子节点，如果插入一个节点，这个原来的节点要往后移动，毕竟你占了ta的位置，那后边如果连着一串是不是都要移动；如果是删除，你后面相邻的节点要补上来，总不能一个空节点挂1个/2个子节点吧，还是要移动大量的节点

**Reply by metahy** [reply]
先说结论：“增删节点需要通过数组插入与删除操作实现，效率较低。”这句话确实有歧义。

如果要通过数组唯一地表示一棵二叉树，那只能采取把二叉树看作**完美二叉树**（推荐）或**完全二叉树**，然后把数组定的足够大预留出足够的空间，这样**当增删节点改变了二叉树高度**的时候，确实涉及到数组的扩容或缩容（缩容可以是非必要的），但如果**增删节点不改变二叉树高度**则直接对数组中对应位置的元素进行覆盖即可。

如果开始把数组容量定的足够大，足够容纳该树最大高度时的所有节点，且内存需求没有紧张到需要根据树的高度缩小数组容量的情况下，这句话就是错误的，这种情况增删节点只需要通过坐标对数组特定位置进行数据修改操作，不影响效率。

**Comment by 0101qwe** [comment]
```python
def dfs(self, i: int, order: str):
        """深度优先遍历"""
        if self.val(i) is None:
            return
        # 前序遍历
        if order == "pre":
            self.res.append(self.val(i))
        self.dfs(self.left(i), order)
        # 中序遍历
        if order == "in":
            self.res.append(self.val(i))
        self.dfs(self.right(i), order)
        # 后序遍历
        if order == "post":
            self.res.append(self.val(i))
```
这里的前，中，后序遍历怎么理解啊。为什么都是先根节点入列表

**Reply by 0101qwe** [reply]
就是这个为什么前序后面递归左子树，不需要递归右子树吗，为什么中序也是先根节点先入，还有后面咋变成右子树递归了，左子树呢，后序也是，为什么直接不用递归了，有点晕，太菜了

**Reply by shinji411** [reply]
因为有左右子树的遍历是和三个if条件平行的，不是上下级关系，注意段前间距，去原文看吧，这里显示不出来

```python
"""深度优先遍历"""
if self.val(i) is None:
    return
// 前序遍历
if order == "pre":
    self.res.append(self.val(i))
self.dfs(self.left(i), order) // 注意此行段前间距
// 中序遍历
if order == "in":
    self.res.append(self.val(i))
self.dfs(self.right(i), order) // 注意此行段前间距
//  后序遍历
if order == "post":
    self.res.append(self.val(i))
```

**Reply by xiaoyaozizaiwoweidu** [reply]
假设：[1, 2, 3, 4, 5, 6, 7]
你可以结合图 7-15   完全二叉树的数组表示的图来看
你理解二叉树的遍历，
理解这个很简单的，你一步一步对着代码和图想，递归递归回溯回溯...
图 7-10   二叉搜索树的前序，中序，后续遍历表示
哪里搞清楚他们的顺序就行了
前序遍历 (中，左，右):
访问节点 1，然后递归地遍历左子树（节点 2），再遍历右子树（节点 3）。
左子树的前序遍历：访问节点 2，然后遍历其左子树（节点 4），再遍历其右子树（节点 5）。
右子树的前序遍历：访问节点 3，然后遍历其左子树（节点 6），再遍历其右子树（节点 7）。
最终结果：[1, 2, 4, 5, 3, 6, 7]
中序遍历 (左，中，右):
递归地遍历左子树（节点 2），然后访问节点 1，再遍历右子树（节点 3）。
左子树的中序遍历：遍历左子树（节点 4），访问节点 2，再遍历右子树（节点 5）。
右子树的中序遍历：遍历左子树（节点 6），访问节点 3，再遍历右子树（节点 7）。
最终结果：[4, 2, 5, 1, 6, 3, 7]
后序遍历 (左，右，中):
递归地遍历左子树（节点 2），然后遍历右子树（节点 3），最后访问节点 1。
左子树的后序遍历：遍历左子树（节点 4），再遍历右子树（节点 5），最后访问节点 2。
右子树的后序遍历：遍历左子树（节点 6），再遍历右子树（节点 7），最后访问节点 3。
最终结果：[4, 5, 2, 6, 7, 3, 1]

**Reply by SenyFish** [reply]
千万不要去想递归函数的执行过程

**Comment by clairetsai1222** [comment]
**深度优先遍历**
```python
def dfs(self, i: int, order: str):
    """深度优先遍历"""
    if self.val(i) is None:
        return
    # 前序遍历
    if order == "pre":
        self.res.append(self.val(i))
        self.dfs(self.left(i), order)
        self.dfs(self.right(i), order)
    # 中序遍历
    if order == "in":
        self.dfs(self.left(i), order)
        self.res.append(self.val(i))
        self.dfs(self.right(i), order)
    # 后序遍历
    if order == "post":
        self.dfs(self.left(i), order)
        self.dfs(self.right(i), order)
        self.res.append(self.val(i))
```

**Comment by xiaozhuge123** [comment]
day06

**Comment by fug-0117** [comment]
代码写的太优雅了

**Comment by yyym2017** [comment]
```java

private void dfs(Integer i, String order, List<Integer> res) {
        // 若为空位，则返回
        if (val(i) == null)
            return;
        // 前序遍历
        if ("pre".equals(order))
            res.add(val(i));
        dfs(left(i), order, res);
        // 中序遍历
        if ("in".equals(order))
            res.add(val(i));
        dfs(right(i), order, res);
        // 后序遍历
        if ("post".equals(order))
            res.add(val(i));
    }

```
这代码第一时间看迷糊了，不得不佩服作者

**Comment by yanlinPeng-code** [comment]
优雅优雅

**Comment by ziventian** [comment]
请教下大家，对于文中的使用数组来表示任意的二叉树的这个示例:
``` 
 # 二叉树的数组表示
 # 使用 None 来表示空位
 tree = [1, 2, 3, 4, None, 6, 7, 8, 9, None, None, 12, None, None, 15]
```

如何才能将图示的二叉树在层序遍历的时候，转换为上面的这个数组，我遍历的时候，转换后总是:
`tree = [1, 2, 3, 4, None, 6, 7, 8, 9, 12, None, None, 15]`

**Reply by metahy** [reply]
你的问题在于：遍历到index为1的节点时，右子节点为空，所有添加了None到数组，当遍历到index为4的这个节点时发现其为None，故而跳过该节点，没有把该None节点的左右子节点None添加进去（没有当作**完全二叉树**来处理）。

这里把None的子节点添加到数组看起来是个挺抽象的行为，但这一步是必不可少的，是为了强制把二叉树凑成**完全二叉树**。所以问题不单是出在层序遍历上，提供的树结构中也要把**非叶节点但值为空**的节点与编程语言中表示空的值（如None或null）区分开来，例如加入新的自定义类型Empty，然后修改层序遍历逻辑：当遇到作为子节点的Empty时，把编程语言的空值加入数组，当遇到作为根节点的Empty时，把其左右子节点均以编程语言的空值加入数组。

最重要的就是理解把二叉树强制凑成**完全二叉树**这一点，沿着这个思路往下走遇到的问题就都好理解和解决了。

**Comment by sssky1** [comment]
```python
       # 如果是前序遍历，左右子树之前访问根
        if order == "pre":
            self.res.append(self.val(i))
        self.dfs(self.left(i), order) # 递归访问左子树
        # 如果是中序遍历，访问左子树之后访问根
        if order == "in":
            self.res.append(self.val(i))
        self.dfs(self.right(i), order) # 递归访问右子树
        # 如果后序遍历，访问左右子树之后访问根
        if order == "post":
            self.res.append(self.val(i))
```
。我觉得注释的不太清楚

**Comment by jyqs9** [comment]
我要学几年才能写上这种代码呢？？？？？？？？？？？？啊啊啊啊啊啊啊啊啊啊啊啊啊啊啊啊啊啊啊

**Comment by guda-art** [comment]
为什么代码里统一写成了 (i - 1) // 2？
你可能会问：“既然左孩子是 (i-1)/2，右孩子是 (i-2)/2，为什么代码里只用 (i - 1) // 2 这一行？”
这是因为代码中使用的是整除（向下取整，//）。让我们看看整除的魔力：
当 i 是左孩子（i 为奇数）时：
例如 i = 3。
真实父节点应该是 (3 - 1) / 2 = 1。
代码计算 (3 - 1) // 2 = 2 // 2 = 1。✅ 正确。
当 i 是右孩子（i 为偶数）时：
例如 i = 4。
真实父节点应该是 (4 - 2) / 2 = 1。
代码计算 (4 - 1) // 2 = 3 // 2 = 1（整除舍去小数）。✅ 也正确。
结论： 利用整数除法向下取整的特性，(i - 1) // 2 这个公式巧妙地同时兼容了左孩子和右孩子的情况。

**Reply by JensenLei123** [reply]
天才，优雅实在是太优雅了

**Comment by Dumbliidore** [comment]
_so phthonic!_
